x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/SamlServiceProviderIndex.java

```diff
@@ -0,0 +1,189 @@
+        client.admin().indices().putTemplate(request, ActionListener.wrap(
+            response -> listener.onResponse(response.isAcknowledged()), listener::onFailure));
+    }
+
+    public void writeDocument(SamlServiceProviderDocument document, ActionListener<String> listener) {
+        final ValidationException exception = document.validate();
+        if (exception != null) {
+            listener.onFailure(exception);
+            return;
+        }
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream();
+             XContentBuilder xContentBuilder = new XContentBuilder(XContentType.JSON.xContent(), out)) {
+            document.toXContent(xContentBuilder, ToXContent.EMPTY_PARAMS);
+            // Due to the lack of "alias templates" (at the current time), we cannot write to the alias if it doesn't exist yet
+            // - that would cause the alias to be created as a concrete index, which is not what we want.
+            // So, until we know that the alias exists we have to write to the expected index name instead.
+            final IndexRequest request = new IndexRequest(aliasExists ? ALIAS_NAME : INDEX_NAME)
+                .source(xContentBuilder)
+                .id(document.docId);
+            client.index(request, ActionListener.wrap(response -> {
+                logger.debug("Wrote service provider [{}][{}] as document [{}]", document.name, document.entityId, response.getId());
+                listener.onResponse(response.getId());
+            }, listener::onFailure));
+        } catch (IOException e) {
+            listener.onFailure(e);
+        }
+    }
+
+    public void readDocument(String documentId, ActionListener<SamlServiceProviderDocument> listener) {
+        final GetRequest request = new GetRequest(ALIAS_NAME, documentId);
+        client.get(request, ActionListener.wrap(response -> {
+            final SamlServiceProviderDocument document = toDocument(documentId, response.getSourceAsBytesRef());
+            listener.onResponse(document);
+        }, listener::onFailure));
+    }
+
+    public void findByEntityId(String entityId, ActionListener<Set<SamlServiceProviderDocument>> listener) {
+        final QueryBuilder query = QueryBuilders.termQuery(SamlServiceProviderDocument.Fields.ENTITY_ID.getPreferredName(), entityId);
+        findDocuments(query, listener);
+    }
+
+    public void findAll(ActionListener<Set<SamlServiceProviderDocument>> listener) {
+        final QueryBuilder query = QueryBuilders.matchAllQuery();
+        findDocuments(query, listener);
+    }
+
+    public void refresh(ActionListener<Void> listener) {
+        client.admin().indices().refresh(new RefreshRequest(ALIAS_NAME), ActionListener.wrap(
+            response -> listener.onResponse(null), listener::onFailure));
+    }
+
+    private void findDocuments(QueryBuilder query, ActionListener<Set<SamlServiceProviderDocument>> listener) {
+        logger.trace("Searching [{}] for [{}]", ALIAS_NAME, query);
```

Comment:
There's not benefit in that. We only need to guard trace/debug messages if generating them is costly. This is a (more-or-less) zero cost call - there's no string addition or object access.

e.g. This:
```
logger.trace("Searching [{}] for [{}]", ALIAS_NAME, convertToString(query));
``` 
is costly, because it calls `convertToString` even if tracing is disabled, so it should either be wrapped in a `Supplier<ParameterizedMessage>` or a `logger.isTraceEnabled()`.
But without that cost, this line will just call `Logger.trace(String, Object[] params)` which will immediately check that trace is disabled and do nothing.